### PR TITLE
DRAFT: Proxy Protocol Transport Socket 

### PR DESCRIPTION
**Pull request by wez470**
I opened this because I wanted to discuss / get some feedback on a few things. I am thinking I can split the rest of this work into 2 PRs (Maybe 3?). Contents would be as follows:

PR 1:
- Transport socket class
- Transport socket unit tests / test utils
- DownstreamAddresses transport socket option struct
- Refactor of listener filter to use common constants

PR 2:
- Transport socket config / config factory
- Transport socket API
- Transport socket Integration tests (Not in this PR yet)
- TCP Proxy API update for option to set down addrs socket option (Not in this PR yet)
- TCP Proxy update to set downstream addr option (Just hacked in for now so I could test things out)
- Any needed TCP Proxy integration / unit tests (Not in this PR yet)
- Documentation (Not in this PR yet)

PR 3:
- Performance pass?

This branch currently has PR 1 and some parts of what would be PR 2 in it. I'd just move the PR 1 pieces to a separate branch. Thoughts? I've also left a comment with a question on some code.

@alyssawilk @mattklein123 

**edit** - Just noticed that v2 xDS is now frozen. So I'll have to update that.

### CODE REVIEW

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    src_address = down_addrs.remote_addr_;
+    dst_address = down_addrs.local_addr_;
+    src_port = down_addrs.remote_port_;
+    dst_port = down_addrs.local_port_;
+  }
+
+  Buffer::OwnedImpl buff{};
+  Common::ProxyProtocol::generateV1Header(src_address, dst_address, src_port, dst_port, ip_version, buff);
+  auto res = buff.write(callbacks_->ioHandle());
+  if (!res.ok()) {
```
Comment (wez470) [inline source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc:102]: Looking at the raw buffer socket, it also checks if the `res.err` is an `Api::IoError::IoErrorCode::Again`, and returns `KeepOpen` if it is. I wasn't sure if when receiving the `Again` error, it was a possibility that the header was partially sent. Handling that would complicate the logic so for now I've opted to return `Close` on any error. Let me know what you think is the right thing to do here is.

File: include/envoy/network/transport_socket.h [hunk 2]
```diff
@@ -184,6 +193,12 @@ class TransportSocketOptions {
    */
   virtual const std::vector<std::string>& applicationProtocolListOverride() const PURE;
 
+  /**
+   * @return the optional downstream address info. Can be used in upstream sockets that
+   *         need awareness of downstream info e.g. proxy_protocol.
+   */
+  virtual const absl::optional<DownstreamAddresses> downstreamAddresses() const PURE;
```
Comment (alyssawilk) [inline include/envoy/network/transport_socket.h:200]: @envoyproxy/maintainers for future-proofing sanity check
We use our own custom proxy proto header equivalent in-house as we add some non-standard fields IMO are fairly generically useful.  Do y'all know if anyone else does this?  I already have some ideas of how we can do our custom stuff with stock Envoy but if we're not the only ones who play around this way I'd be inclined to do a review pass with generic extensibility in mind.

File: include/envoy/network/transport_socket.h [hunk 3]
```diff
@@ -162,6 +163,14 @@ using TransportSocketPtr = std::unique_ptr<TransportSocket>;
  */
 class TransportSocketOptions {
 public:
+  struct DownstreamAddresses {
+    const std::string remote_addr_;
```
[the remark this answers is not in the document]
Comment (htuch) [reply include/envoy/network/transport_socket.h:167]: even worse nit: this is a deviation between Envoy and Google style, we do `_` for structs in Envoy (see https://github.com/envoyproxy/envoy/blob/master/STYLE.md#deviations-from-google-c-style-guidelines).

File: api/envoy/config/transport_socket/proxy_protocol/v2/proxy_protocol.proto [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+option java_outer_classname = "ProxyProtocolProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_migrate).move_to_package =
+    "envoy.extensions.transport_sockets.proxy_protocol.v3";
+
+// [#protodoc-title: Proxy Protocol]
+// PROXY protocol transport socket.
+// [#extension: envoy.transport_sockets.proxy_protocol]
+
+message ProxyProtocol {
```
[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/config/transport_socket/proxy_protocol/v2/proxy_protocol.proto:20]: @alyssawilk are you asking about the version part? If this is actually a transport socket and it's wrapping another, this doesn't seem an unreasonable part for the config.

File: include/envoy/network/transport_socket.h (hunk 2 above)
Comment (ggreenway) [reply include/envoy/network/transport_socket.h:200]: proxy protocol v2 has support to extensions. Can your internal version be mapped on this?

https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt

```
If the length specified in the PROXY protocol header indicates that additional
bytes are part of the header beyond the address information, a receiver may
choose to skip over and ignore those bytes, or attempt to interpret those
bytes.

The information in those bytes will be arranged in Type-Length-Value (TLV
vectors) in the following format.  The first byte is the Type of the vector.
The second two bytes represent the length in bytes of the value (not included
the Type and Length bytes), and following the length field is the number of
bytes specified by the length.

        struct pp2_tlv {
            uint8_t type;
            uint8_t length_hi;
            uint8_t length_lo;
            uint8_t value[0];
        };

A receiver may choose to skip over and ignore the TLVs he is not interested in
or he does not understand. Senders can generate the TLVs only for
the information they choose to publish.

The following types have already been registered for the <type> field :

        #define PP2_TYPE_ALPN           0x01
        #define PP2_TYPE_AUTHORITY      0x02
        #define PP2_TYPE_CRC32C         0x03
        #define PP2_TYPE_NOOP           0x04
        #define PP2_TYPE_UNIQUE_ID      0x05
        #define PP2_TYPE_SSL            0x20
        #define PP2_SUBTYPE_SSL_VERSION 0x21
        #define PP2_SUBTYPE_SSL_CN      0x22
        #define PP2_SUBTYPE_SSL_CIPHER  0x23
        #define PP2_SUBTYPE_SSL_SIG_ALG 0x24
        #define PP2_SUBTYPE_SSL_KEY_ALG 0x25
        #define PP2_TYPE_NETNS          0x30
```

File: api/envoy/config/transport_socket/proxy_protocol/v2/proxy_protocol.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (alyssawilk) [reply api/envoy/config/transport_socket/proxy_protocol/v2/proxy_protocol.proto:20]: No sorry, I'm calling out that we're going to want basically the same proto over in ProxyConfig (added in https://github.com/envoyproxy/envoy/pull/10623/files)

I assume it'd be better to have the proto defined once and used in both places, but I don't know how we feel about this proto being included by the router config, or if it should be placed somewhere else

File: include/envoy/network/transport_socket.h (hunk 2 above)
Comment (alyssawilk) [reply include/envoy/network/transport_socket.h:200]: Possibly, but if other folks are likely to use the extension I think that answers my question - that should be good enough for the common case.  Thanks!

**Comment by alyssawilk** [comment]
Actually can we do one more split, which is trying to land the APIs first?

I've got a WIP patch using your utilities (yay!) and I think I have a handle on API work now:
https://github.com/envoyproxy/envoy/compare/master...alyssawilk:shinkansen_next?expand=1

I think basically it'd be nice to have a standalone message for proxy proto config which has the version, and can optionally be extended later for extensions.  Then in my PR for translating CONNECT I can just pull that whole proto in the ProxyConfig. Alternately I can land the API with my PR, but I figure you've done all the hard work, might as well get the git blame credit :-P

**Comment by wez470** [comment]
Yeah, that sounds good to me. I'm fine if that change happens in your PR. If you need it right away, it's probably best. If it can wait a couple days, I'm happy to do it.

File: api/envoy/config/transport_socket/proxy_protocol/v2/proxy_protocol.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (wez470) [reply api/envoy/config/transport_socket/proxy_protocol/v2/proxy_protocol.proto:20]: Got moved out in this PR: https://github.com/envoyproxy/envoy/pull/10845

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc (hunk 1 above)
Comment (wez470) [reply source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc:102]: @lizan, do you know? Should this do a similar thing to the raw buffer socket, or is it fine to `Close` on any error?

**Comment by wez470** [comment]
Okay, thanks for the initial feedback here, everyone :slightly_smiling_face:. Since there were no objections, I'm going to split out the first PR (as mentioned in the description).

**Comment by wez470** [comment]
Closing as I'm just going to work in making PRs for the individual pieces

**Comment by surki** [comment]
@wez470 wondering if you have other PRs opened, if so please share them

**Comment by wez470** [comment]
Hi, sorry for the delay. I plan to have PR #1 out this weekend.

**Pull request closed without merging**